### PR TITLE
[Handshake] Canonicalize away simple mux+cbranch structures

### LINKED ISSUE #2468: [Handshake] Canonicalize away simple mux+cbranch structures
In the following case, the output of the mux can only ever be the data input to the conditional branch.
image

Canonicalizing this structure results in the removal of a mux, a cbranch, and two fork outputs(!).

### PR DESCRIPTION
**Pull request by RamirezLucas**
This commit adds a new `MuxOp` rewrite pattern that canonicalizes muxes with two data operands originating from the same conditional branch, resulting in the removal of at least one mux and one cbranch. It also adds three new tests to check for the correctness of the rewrite pattern (one "positive", where the rewrite pattern should apply, and two "negatives", where it should not), and updates existing tests to reflect the expected change in canonicalized IR.

Example below (without new rewrite pattern on the left, with on the right).
compare

Fixes #2468

### CODE REVIEW

File: lib/Dialect/Handshake/HandshakeOps.cpp [hunk 1]
```diff
@@ -370,12 +370,43 @@ struct EliminateUnaryMuxesPattern : OpRewritePattern<MuxOp> {
   }
 };
 
+struct EliminateCBranchIntoMuxPattern : OpRewritePattern<MuxOp> {
+  using mlir::OpRewritePattern<MuxOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(MuxOp op,
+                                PatternRewriter &rewriter) const override {
+
+    if (op.getDataOperands().size() != 2)
```
Comment (mortbopet) [inline lib/Dialect/Handshake/HandshakeOps.cpp:378]: Nit: store `op.getDataOperands()` in a variable to avoid calling it 3 times in this function.

**Review by mortbopet — COMMENTED** [review]
Awesome! The rewrite pattern looks good to me.

There are a lot of changes in `test_feedforward.mlir` - I think it makes sense to remove the `--canonicalize` call from that test such that it's only testing one thing in isolation (std-to-handshake) and to avoid these kinds of test updates in the future when more canonicalizers inevitably will be added (cc @Dinistro).

File: lib/Dialect/Handshake/HandshakeOps.cpp (hunk 1 above)
Comment (RamirezLucas) [reply lib/Dialect/Handshake/HandshakeOps.cpp:378]: Will do.

**Comment by RamirezLucas** [comment]
> I think it makes sense to remove the --canonicalize call from that test such that it's only testing one thing in isolation (std-to-handshake) and to avoid these kinds of test updates

Makes sense, should I fix that now or make a separate PR about it?

**Comment by Dinistro** [comment]
The PR looks great, thanks for the work. 

I agree with Morten that we should eventually drop the `--canonicalize` in said test, but this falls under the test cleanup task, I would say: https://github.com/llvm/circt/issues/3668
Thus, I argue that you can keep the PR as is, and we can clean this up later on. Fixing it here would cause even more unrelated changes.

**Review by Dinistro — APPROVED** [review]

**Review by mortbopet — APPROVED** [review]
Fine with me, as long as it gets done!

**Comment by Dinistro** [comment]
> Fine with me, as long as it gets done!

I plan to do some coding next week, so I might tackle some of the long-standing open issues.

**Comment by RamirezLucas** [comment]
> @RamirezLucas should we merge this PR for you?

Yes feel free!

> It might make sense to request commit access to CIRCT as you will probably contribute many more patches. To do so, consult this page: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

Thanks for the link, will look into requesting access.

**Pull request merged**